File: ch8/waitdemo1.c
#include<stdio.h>
#include<unistd.h>
#include<stdlib.h>
#include<sys/types.h>
#include<sys/wait.h>

#define DELAY 2

void parent_code(int);
void child_code(int);

int main(int argc, char const *argv[])
{
    int newpid;

    printf("before: mypid is %d\n",getpid());

    if ((newpid = fork())==-1)
    {
        perror("fork");
    }
    else if (newpid==0)
    {
        child_code(DELAY);
    }
    else
    {
        parent_code(newpid);
    }
    
    return 0;
}

// new process take a nap and then exits
void child_code(int delay){
    printf("child %d here,will sleep %d seconds \n",getpid(),delay);
    sleep(delay);
    printf("child done,about to exit\n");
    exit(17);
}

void parent_code(int chlidpid){
    int wait_rv;
    wait_rv = wait(NULL);
    printf("done waiting for %d. Wait returned %d\n",chlidpid,wait_rv);
}

File: ch8/ch8_src/src/bin/waitdemo1.rs
use nix::sys::wait::{wait, WaitStatus};
use nix::unistd::{fork, ForkResult};
use std::process::exit;
use std::time::Duration;
use std::thread::sleep;

const DELAY: u64 = 2;

fn main() {
    println!("before: mypid is {}", std::process::id());

    match unsafe { fork() } {
        Ok(ForkResult::Parent { child, .. }) => {
            parent_code(child);
        }
        Ok(ForkResult::Child) => {
            child_code(DELAY);
        }
        Err(err) => {
            eprintln!("fork failed: {:?}", err);
        }
    }
}

fn child_code(delay: u64) {
    println!("child {} here, will sleep {} seconds", std::process::id(), delay);
    sleep(Duration::from_secs(delay));
    println!("child done, about to exit");
    exit(17);
}

fn parent_code(_childpid: nix::unistd::Pid) {
    match wait() {
        Ok(WaitStatus::Exited(pid, status)) => {
            println!("done waiting for {}. Wait returned status {}", pid, status);
        }
        Ok(WaitStatus::Signaled(pid, signal, _)) => {
            println!("child {} was terminated by signal {:?}", pid, signal);
        }
        Ok(_) => {
            println!("child exited unexpectedly");
        }
        Err(err) => {
            eprintln!("wait failed: {:?}", err);
        }
    }
}